File: python/ghostcrawl/mcp/wrapper.py
"""Typed MCP client wrapping mcp.ClientSession over streamable-HTTP transport.

Hand-written (~200 LOC); a thin typed layer over the MCP tool surface, distinct
from the REST client.

Transport: ``mcp.client.streamable_http.streamable_http_client``
(the current name; the old non-underscore alias is deprecated).

Every method below corresponds to a registered MCP tool on the GhostCrawl
server. Methods where the tool schema requires a primary positional argument
expose that argument explicitly; all other parameters flow through ``**kwargs``.
"""

from __future__ import annotations

from contextlib import asynccontextmanager
from typing import Any, AsyncGenerator

import httpx
from mcp import ClientSession
from mcp.client.streamable_http import streamable_http_client  # current name; old alias is deprecated


# ----------------------------------------------------------------------------
# Canonical tool-name reconciliation (D-11)
# ----------------------------------------------------------------------------
# The native GhostCrawl MCP server (``ghostcrawl.mcp.tools.TOOL_REGISTRY``) exposes
# every tool under the canonical ``ghostcrawl_*`` (data-plane) / ``act_*`` (browser
# primitive) / ``action_record_*`` namespaces — NOT the bare ``navigate`` / ``act``
# aliases the standalone (now-retired) server used. This wrapper keeps ergonomic
# method names but reconciles each call to the real registry tool name before it
# hits the wire, so it drives the native streamable-HTTP server directly.
_TOOL_NAMES: dict[str, str] = {
    "navigate": "ghostcrawl_navigate",
    "act": "ghostcrawl_act",
    "act_on": "ghostcrawl_act",
    "observe": "ghostcrawl_observe",
    "extract": "ghostcrawl_extract",
    "scrape": "ghostcrawl_scrape",
    "screenshot": "ghostcrawl_screenshot",
    "screenshot_of": "ghostcrawl_screenshot",
    "search": "ghostcrawl_search",
    "google_search": "ghostcrawl_search",
    "map_site": "ghostcrawl_map",
    "pdf": "ghostcrawl_pdf",
    "crawl": "ghostcrawl_crawl",
    "eval": "act_evaluate",
    "run_script": "ghostcrawl_script_run",
    "upload_file": "act_upload",
    "start": "ghostcrawl_session_create",
    "end": "ghostcrawl_session_terminate",
    "start_recording": "action_record_start",
    "stop_recording": "action_record_stop",
    "ping": "ping",  # server health tool is registered literally as ``ping``
}

_CANONICAL_PREFIXES = ("ghostcrawl_", "act_", "action_")


def _canonical_tool(name: str) -> str:
    """Map a wrapper method's logical name to the real MCP registry tool name.

    Explicit mappings win; anything already in a canonical namespace passes
    through unchanged; everything else takes the ``ghostcrawl_`` data-plane
    prefix (the naming rule the server enforces for GhostCrawl-owned tools).
    """
    if name in _TOOL_NAMES:
        return _TOOL_NAMES[name]
    if name.startswith(_CANONICAL_PREFIXES) or name == "ping":
        return name
    return f"ghostcrawl_{name}"


class GhostCrawlMCPClient:
    """Typed wrapper for the ghostcrawl MCP tool surface via streamable-HTTP.

    Parameters
    ----------
    mcp_url:
        Full URL of the MCP endpoint, e.g. ``http://localhost:8090/mcp``.
    api_key:
        Bearer token for the API.  Read from ``GHOSTCRAWL_API_KEY`` env var
        in CLI contexts; passed explicitly when constructing programmatically.
    timeout:
        Per-request timeout in seconds (default: 30).
    """

    def __init__(self, mcp_url: str, api_key: str, timeout: float = 30.0) -> None:
        if not mcp_url:
            raise ValueError("mcp_url is required and must not be empty")
        if not api_key:
            raise ValueError("api_key is required and must not be empty")
        self._url = mcp_url
        self._api_key = api_key
        self._timeout = timeout
        self._headers: dict[str, str] = {"Authorization": f"Bearer {api_key}"}

    # ------------------------------------------------------------------
    # Session context manager
    # ------------------------------------------------------------------

    @asynccontextmanager
    async def session(self) -> AsyncGenerator[ClientSession, None]:
        """Open an MCP session over streamable-HTTP and yield it.

        Callers should use this when they want to batch multiple tool calls
        in a single connection.  All public tool methods open their own
        short-lived sessions via this context manager internally.
        """
        http = httpx.AsyncClient(headers=self._headers, timeout=self._timeout)
        try:
            async with streamable_http_client(self._url, http_client=http) as (
                read,
                write,
                _get_session_id,
            ):
                async with ClientSession(read, write) as s:
                    await s.initialize()
                    yield s
        finally:
            await http.aclose()

    # ------------------------------------------------------------------
    # Internal helper
    # ------------------------------------------------------------------

    async def _call(self, tool_name: str, arguments: dict[str, Any]) -> Any:
        """Open a session, call *tool_name* with *arguments*, and return the result.

        *tool_name* is reconciled to the canonical native-registry name
        (``ghostcrawl_*`` / ``act_*``) via :func:`_canonical_tool` before the call.
        """
        async with self.session() as s:
            return await s.call_tool(_canonical_tool(tool_name), arguments)

    # ------------------------------------------------------------------
    # Browser / navigation tools
    # ------------------------------------------------------------------

    async def navigate(self, url: str, **kwargs: Any) -> Any:
        """Navigate the current browser context to *url*."""
        return await self._call("navigate", {"url": url, **kwargs})

    async def act(self, goal: str, **kwargs: Any) -> Any:
        """Perform a high-level browser action described by *goal*."""
        return await self._call("act", {"goal": goal, **kwargs})

    async def observe(self, **kwargs: Any) -> Any:
        """Return an observation of the current browser state."""
        return await self._call("observe", kwargs)

    async def extract(self, **kwargs: Any) -> Any:
        """Extract structured data from the current page."""
        return await self._call("extract", kwargs)

    async def scrape(self, url: str, **kwargs: Any) -> Any:
        """Scrape content from *url* and return extracted data."""
        return await self._call("scrape", {"url": url, **kwargs})

    async def screenshot(self, **kwargs: Any) -> Any:
        """Take a screenshot of the current page or a given URL."""
        return await self._call("screenshot", kwargs)

    async def screenshot_of(self, selector: str, **kwargs: Any) -> Any:
        """Take a screenshot of the element matching *selector*."""
        return await self._call("screenshot_of", {"selector": selector, **kwargs})

    async def dom_snapshot(self, **kwargs: Any) -> Any:
        """Return a DOM snapshot of the current page."""
        return await self._call("dom_snapshot", kwargs)

    async def network(self, **kwargs: Any) -> Any:
        """Retrieve network request/response data for the current page."""
        return await self._call("network", kwargs)

    async def network_har(self, **kwargs: Any) -> Any:
        """Return a HAR archive of network traffic for the current page."""
        return await self._call("network_har", kwargs)

    async def scroll(self, **kwargs: Any) -> Any:
        """Scroll the page or a specific element."""
        return await self._call("scroll", kwargs)

    async def cookies(self, **kwargs: Any) -> Any:
        """Get or set cookies for the current browser context."""
        return await self._call("cookies", kwargs)

    async def wait(self, **kwargs: Any) -> Any:
        """Wait for a condition (selector, network idle, timeout, etc.)."""
        return await self._call("wait", kwargs)

    async def eval(self, **kwargs: Any) -> Any:
        """Evaluate arbitrary JavaScript in the current page context."""
        return await self._call("eval", kwargs)

    async def run_script(self, **kwargs: Any) -> Any:
        """Run a pre-defined browser script by name."""
        return await self._call("run_script", kwargs)

    async def act_on(self, **kwargs: Any) -> Any:
        """Perform an action on a specific element."""
        return await self._call("act_on", kwargs)

    # ------------------------------------------------------------------
    # File / storage tools
    # ------------------------------------------------------------------

    async def upload_file(self, **kwargs: Any) -> Any:
        """Upload a file to the current page context."""
        return await self._call("upload_file", kwargs)

    async def download_file(self, **kwargs: Any) -> Any:
        """Download a file from the current page."""
        return await self._call("download_file", kwargs)

    async def storage_state(self, **kwargs: Any) -> Any:
        """Get or set the storage state (cookies, localStorage) of the context."""
        return await self._call("storage_state", kwargs)

    # ------------------------------------------------------------------
    # Crawl / dataset tools
    # ------------------------------------------------------------------

    async def crawl(self, **kwargs: Any) -> Any:
        """Start or manage a crawl job."""
        return await self._call("crawl", kwargs)

    async def dataset(self, **kwargs: Any) -> Any:
        """Interact with a ghostcrawl dataset resource."""
        return await self._call("dataset", kwargs)

    # ------------------------------------------------------------------
    # Session lifecycle tools
    # ------------------------------------------------------------------

    async def start(self, **kwargs: Any) -> Any:
        """Start a new browser session."""
        return await self._call("start", kwargs)

    async def end(self, **kwargs: Any) -> Any:
        """End an existing browser session."""
        return await self._call("end", kwargs)

    async def start_many(self, **kwargs: Any) -> Any:
        """Start multiple browser sessions at once."""
        return await self._call("start_many", kwargs)

    async def end_all(self, **kwargs: Any) -> Any:
        """End all active browser sessions."""
        return await self._call("end_all", kwargs)

    # ------------------------------------------------------------------
    # Recording tools
    # ------------------------------------------------------------------

    async def start_recording(self, **kwargs: Any) -> Any:
        """Start recording a browser session."""
        return await self._call("start_recording", kwargs)

    async def stop_recording(self, **kwargs: Any) -> Any:
        """Stop the current recording."""
        return await self._call("stop_recording", kwargs)

    async def list_recordings(self, **kwargs: Any) -> Any:
        """List all stored recordings."""
        return await self._call("list_recordings", kwargs)

    async def get_recording(self, **kwargs: Any) -> Any:
        """Retrieve a specific recording by ID."""
        return await self._call("get_recording", kwargs)

    async def delete_recording(self, **kwargs: Any) -> Any:
        """Delete a recording by ID."""
        return await self._call("delete_recording", kwargs)

    async def ping(self) -> Any:
        """Health-check tool.  Returns 'pong'.  Verifies server is reachable.

        Mirrors the server-side ``@mcp.tool() ping()`` registered in
        ``ghostcrawl_mcp_server/server.py``.
        """
        return await self._call("ping", {})

    # ------------------------------------------------------------------
    # Search tools
    # ------------------------------------------------------------------

    async def search(self, query: str, **kwargs: Any) -> Any:
        """Search the web and return structured results."""
        return await self._call("search", {"query": query, **kwargs})

    async def google_search(self, query: str, **kwargs: Any) -> Any:
        """Search Google (or a specific Google vertical) and return SERP results."""
        return await self._call("google_search", {"query": query, **kwargs})

    # ------------------------------------------------------------------
    # Site map tools
    # ------------------------------------------------------------------

    async def map_site(self, url: str, **kwargs: Any) -> Any:
        """Map all pages on a website starting from *url*."""
        return await self._call("map_site", {"url": url, **kwargs})

    async def discovery(self, domain: str, **kwargs: Any) -> Any:
        """Retrieve previously discovered links for *domain*."""
        return await self._call("discovery", {"domain": domain, **kwargs})

    # ------------------------------------------------------------------
    # Identity / profile tools
    # ------------------------------------------------------------------

    async def identity(self, **kwargs: Any) -> Any:
        """Materialise a fresh browser identity."""
        return await self._call("identity", kwargs)

    async def profiles(self, **kwargs: Any) -> Any:
        """Create, get, list, delete, or view recent browser profiles."""
        return await self._call("profiles", kwargs)

    # ------------------------------------------------------------------
    # PDF tool
    # ------------------------------------------------------------------

    async def pdf(self, url: str, **kwargs: Any) -> Any:
        """Render a web page at *url* to PDF."""
        return await self._call("pdf", {"url": url, **kwargs})

    # ------------------------------------------------------------------
    # Key-value store
    # ------------------------------------------------------------------

    async def kv(self, **kwargs: Any) -> Any:
        """Read, write, or delete entries in the persistent key-value store."""
        return await self._call("kv", kwargs)

    # ------------------------------------------------------------------
    # Schedule tools
    # ------------------------------------------------------------------

    async def schedule(self, **kwargs: Any) -> Any:
        """Create, list, get, update, or delete recurring schedules."""
        return await self._call("schedule", kwargs)

    # ------------------------------------------------------------------
    # Webhook tools
    # ------------------------------------------------------------------

    async def webhook(self, **kwargs: Any) -> Any:
        """Register, list, inspect, retry, or rotate secrets for webhooks."""
        return await self._call("webhook", kwargs)

    # ------------------------------------------------------------------
    # Budget / usage tools
    # ------------------------------------------------------------------

    async def budget(self, **kwargs: Any) -> Any:
        """Manage spending controls and view usage for your account."""
        return await self._call("budget", kwargs)

    # ------------------------------------------------------------------
    # Unblock tool
    # ------------------------------------------------------------------

    async def unblock(self, url: str, **kwargs: Any) -> Any:
        """Retrieve *url* reliably through the managed browser fleet."""
        return await self._call("unblock", {"url": url, **kwargs})

    # ------------------------------------------------------------------
    # Lighthouse tool
    # ------------------------------------------------------------------

    async def lighthouse(self, url: str, **kwargs: Any) -> Any:
        """Run a performance and quality audit on *url*."""
        return await self._call("lighthouse", {"url": url, **kwargs})

    # ------------------------------------------------------------------
    # Queue tools
    # ------------------------------------------------------------------

    async def queue(self, **kwargs: Any) -> Any:
        """Push, pop, acknowledge, or view stats for a named task queue."""
        return await self._call("queue", kwargs)

    # ------------------------------------------------------------------
    # Self-host / account utilities
    # ------------------------------------------------------------------

    async def check_updates(self, **kwargs: Any) -> Any:
        """Check for available platform updates."""
        return await self._call("check_updates", kwargs)

    async def registry_token(self, **kwargs: Any) -> Any:
        """Obtain a short-lived container image pull token."""
        return await self._call("registry_token", kwargs)

    # ------------------------------------------------------------------
    # Takeover tools
    # ------------------------------------------------------------------

    async def takeover_token(self, session_id: str, **kwargs: Any) -> Any:
        """Mint a short-lived token to take live control of *session_id*."""
        return await self._call("takeover_token", {"session_id": session_id, **kwargs})

    async def takeover_release(self, session_id: str, **kwargs: Any) -> Any:
        """Release a live-view token for *session_id*."""
        return await self._call("takeover_release", {"session_id": session_id, **kwargs})

File: node/src/mcp/wrapper.ts
/**
 * GhostCrawlMCPClient — hand-written MCP wrapper.
 *
 * A standalone MCP transport client (independent of the @ghostcrawl/sdk REST
 * client). Provides typed async methods over the native GhostCrawl MCP server's
 * tool surface (ghostcrawl.mcp.tools.TOOL_REGISTRY — the ghostcrawl_* data-plane
 * and act_* browser-primitive tools), reconciled to the canonical registry names
 * by canonicalTool() below. This is an OPTIONAL convenience client — any standard
 * MCP client can drive the same hosted server.
 *
 * Transport: StreamableHTTPClientTransport (current; replaces deprecated SSE).
 * Security: API key passed as Bearer token in Authorization header only.
 *   No --api-key CLI arg, no process.argv exposure.
 *
 * Return type is Promise<unknown> — MCP server tool output schemas are absent
 * from openapi/v1.json (Q2). `unknown` forces callers to narrow, satisfying
 * Typed interfaces, not any.
 *
 * Each call opens a fresh MCP client connection and closes it in finally{}.
 * This matches the Python wrapper's per-call session pattern.
 */

import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { StreamableHTTPClientTransport } from "@modelcontextprotocol/sdk/client/streamableHttp.js";

// ─── Canonical tool-name reconciliation (D-11) ──────────────────────────────
// The native GhostCrawl MCP server (ghostcrawl.mcp.tools.TOOL_REGISTRY) exposes
// every tool under the canonical `ghostcrawl_*` (data-plane) / `act_*` (browser
// primitive) / `action_record_*` namespaces — NOT the bare `navigate` / `act`
// aliases the retired standalone server used. This wrapper keeps ergonomic method
// names but reconciles each call to the real registry tool name before it hits the
// wire, so it drives the native streamable-HTTP server directly.
const TOOL_NAMES: Record<string, string> = {
  navigate: "ghostcrawl_navigate",
  act: "ghostcrawl_act",
  act_on: "ghostcrawl_act",
  observe: "ghostcrawl_observe",
  extract: "ghostcrawl_extract",
  scrape: "ghostcrawl_scrape",
  screenshot: "ghostcrawl_screenshot",
  screenshot_of: "ghostcrawl_screenshot",
  search: "ghostcrawl_search",
  google_search: "ghostcrawl_search",
  map_site: "ghostcrawl_map",
  pdf: "ghostcrawl_pdf",
  crawl: "ghostcrawl_crawl",
  eval: "act_evaluate",
  run_script: "ghostcrawl_script_run",
  upload_file: "act_upload",
  start: "ghostcrawl_session_create",
  end: "ghostcrawl_session_terminate",
  start_recording: "action_record_start",
  stop_recording: "action_record_stop",
  ping: "ping", // server health tool is registered literally as `ping`
};

const CANONICAL_PREFIXES = ["ghostcrawl_", "act_", "action_"];

/** Map a wrapper method's logical name to the real MCP registry tool name. */
function canonicalTool(name: string): string {
  if (name in TOOL_NAMES) return TOOL_NAMES[name];
  if (name === "ping" || CANONICAL_PREFIXES.some((p) => name.startsWith(p))) return name;
  return `ghostcrawl_${name}`;
}

/** Options accepted by GhostCrawlMCPClient constructor. */
export interface GhostCrawlMCPClientOptions {
  /** Full URL to the MCP HTTP endpoint, e.g. http://localhost:8090/mcp */
  mcpUrl: string;
  /** GhostCrawl API key — sourced from GHOSTCRAWL_API_KEY env var by callers */
  apiKey: string;
}

export class GhostCrawlMCPClient {
  private readonly mcpUrl: string;
  private readonly apiKey: string;

  constructor(options: GhostCrawlMCPClientOptions) {
    const { mcpUrl, apiKey } = options;
    if (!mcpUrl || typeof mcpUrl !== "string") {
      throw new TypeError("mcpUrl is required and must be a non-empty string");
    }
    if (!apiKey || typeof apiKey !== "string") {
      throw new TypeError("apiKey is required and must be a non-empty string");
    }
    this.mcpUrl = mcpUrl;
    this.apiKey = apiKey;
  }

  /** Create a connected MCP Client for a single tool call. */
  private async createClient(): Promise<Client> {
    const transport = new StreamableHTTPClientTransport(new URL(this.mcpUrl), {
      requestInit: {
        headers: {
          Authorization: `Bearer ${this.apiKey}`,
        },
      },
    });
    const client = new Client({ name: "ghostcrawl-sdk", version: "0.2.0" });
    await client.connect(transport);
    return client;
  }

  /** Invoke a named MCP tool with arguments. Handles connect + close lifecycle. */
  private async callTool(
    name: string,
    args: Record<string, unknown>
  ): Promise<unknown> {
    const client = await this.createClient();
    try {
      // Reconcile to the canonical native-registry tool name (ghostcrawl_*/act_*).
      return await client.callTool({ name: canonicalTool(name), arguments: args });
    } finally {
      await client.close();
    }
  }

  // ─── Navigation & Interaction Tools ────────────────────────────────────────

  /** Navigate the browser to a URL. Required: url. */
  async navigate(params: { url: string; [key: string]: unknown }): Promise<unknown> {
    return this.callTool("navigate", params);
  }

  /** Perform an action described in natural language. Required: goal. */
  async act(params: { goal: string; [key: string]: unknown }): Promise<unknown> {
    return this.callTool("act", params);
  }

  /** Observe the current page state and return structured data. */
  async observe(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.callTool("observe", params);
  }

  /** Extract structured data from the current page. */
  async extract(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.callTool("extract", params);
  }

  // ─── Page Capture Tools ─────────────────────────────────────────────────────

  /** Scrape a URL and return content. Required: url. */
  async scrape(params: { url: string; [key: string]: unknown }): Promise<unknown> {
    return this.callTool("scrape", params);
  }

  /** Take a screenshot of the page or a URL. */
  async screenshot(params: { url?: string; [key: string]: unknown } = {}): Promise<unknown> {
    return this.callTool("screenshot", params);
  }

  /** Take a screenshot of a specific DOM selector. Required: selector. */
  async screenshot_of(params: { selector: string; [key: string]: unknown }): Promise<unknown> {
    return this.callTool("screenshot_of", params);
  }

  /** Perform an action on a specific element. Required: selector, goal. */
  async act_on(params: {
    selector: string;
    goal: string;
    [key: string]: unknown;
  }): Promise<unknown> {
    return this.callTool("act_on", params);
  }

  /** Capture network activity on the current page. */
  async network(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.callTool("network", params);
  }

  /** Capture a DOM snapshot of the current page. */
  async dom_snapshot(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.callTool("dom_snapshot", params);
  }

  /** Scroll on the current page. */
  async scroll(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.callTool("scroll", params);
  }

  /** Wait for a condition or duration on the current page. */
  async wait(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.callTool("wait", params);
  }

  /** Evaluate JavaScript on the current page. */
  async eval(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.callTool("eval", params);
  }

  /** Retrieve cookies from the current page context. */
  async cookies(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.callTool("cookies", params);
  }

  // ─── Crawl Tools ────────────────────────────────────────────────────────────

  /** Start a crawl operation. */
  async crawl(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.callTool("crawl", params);
  }

  // ─── Dataset Tools ──────────────────────────────────────────────────────────

  /** Interact with a dataset resource. */
  async dataset(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.callTool("dataset", params);
  }

  // ─── Script Tools ───────────────────────────────────────────────────────────

  /** Execute a server-side script. */
  async run_script(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.callTool("run_script", params);
  }

  // ─── Session Lifecycle Tools ─────────────────────────────────────────────────

  /** Start a new browser session. */
  async start(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.callTool("start", params);
  }

  /** End an active browser session. */
  async end(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.callTool("end", params);
  }

  /** Start multiple browser sessions. */
  async start_many(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.callTool("start_many", params);
  }

  /** End all active browser sessions. */
  async end_all(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.callTool("end_all", params);
  }

  // ─── File I/O Tools ──────────────────────────────────────────────────────────

  /** Upload a file to the browser session. */
  async upload_file(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.callTool("upload_file", params);
  }

  /** Download a file from the browser session. */
  async download_file(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.callTool("download_file", params);
  }

  // ─── Network & Storage Tools ─────────────────────────────────────────────────

  /** Capture HAR network trace. */
  async network_har(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.callTool("network_har", params);
  }

  /** Get or set browser storage state. */
  async storage_state(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.callTool("storage_state", params);
  }

  // ─── Recording Tools ─────────────────────────────────────────────────────────

  /** List available recordings. */
  async list_recordings(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.callTool("list_recordings", params);
  }

  /** Get a specific recording. */
  async get_recording(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.callTool("get_recording", params);
  }

  /** Start a new recording. */
  async start_recording(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.callTool("start_recording", params);
  }

  /** Stop an active recording. */
  async stop_recording(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.callTool("stop_recording", params);
  }

  /** Delete a recording. */
  async delete_recording(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.callTool("delete_recording", params);
  }

  /**
   * Health-check tool.  Returns 'pong'.  Verifies server is reachable.
   * Mirrors the server-side @mcp.tool() ping() in ghostcrawl_mcp_server/server.py.
   */
  async ping(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.callTool("ping", params);
  }

  // ─── Search Tools ────────────────────────────────────────────

  /** Search the web and return structured results. Required: query. */
  async search(params: { query: string; [key: string]: unknown }): Promise<unknown> {
    return this.callTool("search", params);
  }

  /** Search Google or a Google vertical and return SERP results. Required: query. */
  async google_search(params: { query: string; [key: string]: unknown }): Promise<unknown> {
    return this.callTool("google_search", params);
  }

  // ─── Site Map Tools ───────────────────────────────────────────

  /** Map all pages on a website starting from a URL. Required: url. */
  async map_site(params: { url: string; [key: string]: unknown }): Promise<unknown> {
    return this.callTool("map_site", params);
  }

  /** Retrieve previously discovered links for a domain. Required: domain. */
  async discovery(params: { domain: string; [key: string]: unknown }): Promise<unknown> {
    return this.callTool("discovery", params);
  }

  // ─── Identity / Profile Tools ─────────────────────────────────

  /** Materialise a fresh browser identity. */
  async identity(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.callTool("identity", params);
  }

  /** Create, get, list, delete, or view recent browser profiles. */
  async profiles(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.callTool("profiles", params);
  }

  // ─── PDF Tool ─────────────────────────────────────────────────

  /** Render a web page to PDF. Required: url. */
  async pdf(params: { url: string; [key: string]: unknown }): Promise<unknown> {
    return this.callTool("pdf", params);
  }

  // ─── Key-Value Store ──────────────────────────────────────────

  /** Read, write, or delete entries in the persistent key-value store. */
  async kv(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.callTool("kv", params);
  }

  // ─── Schedule Tools ───────────────────────────────────────────

  /** Create, list, get, update, or delete recurring schedules. */
  async schedule(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.callTool("schedule", params);
  }

  // ─── Webhook Tools ────────────────────────────────────────────

  /** Register, list, inspect, retry, or rotate secrets for webhooks. */
  async webhook(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.callTool("webhook", params);
  }

  // ─── Budget / Usage Tools ─────────────────────────────────────

  /** Manage spending controls and view usage for the account. */
  async budget(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.callTool("budget", params);
  }

  // ─── Unblock Tool ─────────────────────────────────────────────

  /** Retrieve a URL reliably through the managed browser fleet. Required: url. */
  async unblock(params: { url: string; [key: string]: unknown }): Promise<unknown> {
    return this.callTool("unblock", params);
  }

  // ─── Lighthouse Tool ──────────────────────────────────────────

  /** Run a performance and quality audit on a web page. Required: url. */
  async lighthouse(params: { url: string; [key: string]: unknown }): Promise<unknown> {
    return this.callTool("lighthouse", params);
  }

  // ─── Queue Tools ──────────────────────────────────────────────

  /** Push, pop, acknowledge, or view stats for a named task queue. */
  async queue(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.callTool("queue", params);
  }

  // ─── Self-host / Account Utilities ────────────────────────────

  /** Check for available platform updates. */
  async check_updates(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.callTool("check_updates", params);
  }

  /** Obtain a short-lived container image pull token. */
  async registry_token(params: Record<string, unknown> = {}): Promise<unknown> {
    return this.callTool("registry_token", params);
  }

  // ─── Takeover Tools ───────────────────────────────────────────

  /** Mint a short-lived token to take live control of a session. Required: session_id. */
  async takeover_token(params: { session_id: string; [key: string]: unknown }): Promise<unknown> {
    return this.callTool("takeover_token", params);
  }

  /** Release a live-view token for a session. Required: session_id. */
  async takeover_release(params: { session_id: string; [key: string]: unknown }): Promise<unknown> {
    return this.callTool("takeover_release", params);
  }
}
